dask_cuml/linear_model/linear_regression.py

```diff
@@ -0,0 +1,511 @@
+                                                  part, worker, loc_dict,
+                                                  X_df.npartitions,
+                                                  workers=[worker]))
+                           for worker, part in worker_parts.items()]
+
+        yield wait(input_devarrays)
+
+        """
+        Gather IPC handles for each worker and call _fit() on each worker
+        containing data.
+        """
+        exec_node = input_devarrays[0][0]
+
+        # Need to fetch parts on worker
+        on_worker = list(filter(lambda x: x[0] == exec_node, input_devarrays))
+        not_on_worker = list(filter(lambda x: x[0] != exec_node,
+                                    input_devarrays))
+
+        ipc_handles = [client.submit(get_input_ipc_handles, future,
+                                     workers=[a_worker])
+                       for a_worker, future in not_on_worker]
+
+        raw_arrays = [future for a_worker, future in on_worker]
+
+        # IPC Handles are loaded in separate threads on worker so they can be
+        # used to make calls through cython
+        # Calls _predict_on_worker defined in the bottom
+        ret = client.submit(_predict_on_worker, (ipc_handles, raw_arrays),
+                            self._build_params_map(), workers=[exec_node])
+
+        yield wait(ret)
+
+        dfs = [client.submit(series_on_worker, f, worker, loc_dict,
+                             X_df.npartitions, X_df, workers=[worker])
+               for worker, f in input_devarrays]
+
+        return dfs
+
+    def _build_host_dict(self, gpu_futures, client):
+
+        who_has = client.who_has(gpu_futures)
+
+        key_to_host_dict = {}
+        for key in who_has:
+            key_to_host_dict[key] = parse_host_port(who_has[key][0])
+
+        hosts_to_key_dict = {}
+        for key, host in key_to_host_dict.items():
+            if host not in hosts_to_key_dict:
+                hosts_to_key_dict[host] = set([key])
+            else:
+                hosts_to_key_dict[host].add(key)
+
+        workers = [key[0] for key in list(who_has.values())]
+        return build_host_dict(workers)
+
+
+def _fit_on_worker(data, params):
+    ipc_dev_list, devarrs_dev_list = data
+
+    # TODO: One ipc thread per device instead of per x,y,coef tuple
+    open_ipcs = []
+    for p, dev in ipc_dev_list:
+        for x, y, coef in p:
+            ipct = new_ipc_thread([x, y, coef], dev)
+            open_ipcs.append(ipct)
+
+    alloc_info = list(itertools.chain([t.info() for t in open_ipcs]))
+    alloc_info.extend(
+        list(itertools.chain(
+            [[(build_alloc_info(X)[0], build_alloc_info(y)[0],
+               build_alloc_info(coef)[0]) for X, y, coef in p]
+             for p, dev in devarrs_dev_list])))
+
+    print("ALLOC INFO FIT")
+    print(alloc_info)
+
+    try:
+        ols = cuOLS()
+
+        intercept = ols.fit_dask(alloc_info, params)
```

Comment:
I was thinking we were trying to keep all dask related verbiage out of our cuml API to keep these functions portable (and decoupled from any specific distributed scheduler). We never discussed this, I just kind of assumed it. What are your thoughts on this?

Reply:
I totally agree, my in line comment on the cuml side indeed agrees: `#todo: cleanup in dask and here to avoid so much subindexing and not depend on exact alloc_info object`